### PR TITLE
MM-45994 ensure database operations return their errors

### PR DESCRIPTION
**Pull request by noxer**
#### Summary
This PR adds a number of returns for previously ignored errors in the sqlstore package. It also adds functionality to report errors in the rollback of operations. In once instance, it removes the reporting of errors returned by `bytes.Buffer`, which only panics but never actually returns errors.

#### Ticket Link
[MM-45994](https://mattermost.atlassian.net/browse/MM-45994)

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: store/sqlstore/adapters.go [hunk 1]
```diff
@@ -17,23 +17,18 @@ type jsonArray []string
 
 func (a jsonArray) Value() (driver.Value, error) {
 	var out bytes.Buffer
-	if err := out.WriteByte('['); err != nil {
-		return nil, err
-	}
+	_ = out.WriteByte('[')
```
Comment (agnivade) [inline store/sqlstore/adapters.go:20]: Hmm .. what is the reason for these? I can see that this won't return an error, but the code seemed idiomatic.

File: store/sqlstore/channel_store.go [hunk 2]
```diff
@@ -402,6 +402,7 @@ func (db allChannelMember) Process() (string, string) {
 		for _, role := range roles {
 			if role == impliedRole {
 				alreadyThere = true
+				break
```
Comment (agnivade) [inline store/sqlstore/channel_store.go:405]: nice!

File: store/sqlstore/channel_store.go [hunk 3]
```diff
@@ -1563,8 +1564,7 @@ func (s SqlChannelStore) getByName(teamId string, name string, includeDeleted bo
 		return nil, errors.Wrapf(err, "failed to find channel with TeamId=%s and Name=%s", teamId, name)
 	}
 
-	channelByNameCache.SetWithExpiry(teamId+name, &channel, ChannelCacheDuration)
-	return &channel, nil
+	return &channel, channelByNameCache.SetWithExpiry(teamId+name, &channel, ChannelCacheDuration)
```
Comment (agnivade) [inline store/sqlstore/channel_store.go:1567]: Let's revert this change. This is making it slightly harder to read.

File: store/sqlstore/scheme_store.go [hunk 4]
```diff
@@ -22,20 +22,20 @@ func newSqlSchemeStore(sqlStore *SqlStore) store.SchemeStore {
 	return &SqlSchemeStore{sqlStore}
 }
 
-func (s *SqlSchemeStore) Save(scheme *model.Scheme) (*model.Scheme, error) {
+func (s *SqlSchemeStore) Save(scheme *model.Scheme) (_ *model.Scheme, err error) {
 	if scheme.Id == "" {
-		transaction, err := s.GetMasterX().Beginx()
-		if err != nil {
-			return nil, errors.Wrap(err, "begin_transaction")
+		transaction, terr := s.GetMasterX().Beginx()
```
Comment (agnivade) [inline store/sqlstore/scheme_store.go:27]: Wondering why is this `terr` and not just `err`?

**Review by agnivade — COMMENTED** [review]
Very thorough work! Just some small comments.

**Review by isacikgoz — APPROVED** [review]
LGTM 👍

File: store/sqlstore/adapters.go (hunk 1 above)
Comment (noxer) [reply store/sqlstore/adapters.go:20]: It was a lot of lines that did nothing. Why not get rid of them for simplicity?

File: store/sqlstore/channel_store.go (hunk 2 above)
Comment (noxer) [reply store/sqlstore/channel_store.go:405]: I couldn't get past this one, sorry :D

File: store/sqlstore/channel_store.go (hunk 3 above)
Comment (noxer) [reply store/sqlstore/channel_store.go:1567]: Revert it or assign it to `err` and return that?

File: store/sqlstore/scheme_store.go (hunk 4 above)
Comment (noxer) [reply store/sqlstore/scheme_store.go:27]: Because otherwise we'd shadow `err` which triggers vet.

File: store/sqlstore/channel_store.go (hunk 3 above)
Comment (agnivade) [reply store/sqlstore/channel_store.go:1567]: Assign to `err` and return :)

File: store/sqlstore/scheme_store.go (hunk 4 above)
Comment (agnivade) [reply store/sqlstore/scheme_store.go:27]: Oh yes, we usually use `err2` for cases like that. But this is okay as well.

File: store/sqlstore/adapters.go (hunk 1 above)
Comment (agnivade) [reply store/sqlstore/adapters.go:20]: It just makes to code simpler to read. Having `_` leads to the question "why are we ignoring the errors?". Then the user has to realize that it is being ignored deliberately because it is guaranteed to not return errors. I think the cost of this mental complexity is higher than saving a couple of lines of code.

**Review by agnivade — APPROVED** [review]
Excellent work!

**Comment by noxer** [comment]
Looks good
